docs/development.md

```diff
@@ -60,4 +60,44 @@ To generate `zz_generated.deepcopy.go` & `zz_generated.swagger_doc_generated.go`
 3. Run `make update-scripts` to see `zz_generated.deepcopy.go` & `zz_generated.swagger_doc_generated.go` generated in your API directory, and along with some related scripts generated in `client` directory. If `zz_generated.deepcopy.go` is not generated properly, please run the command in a container with `RUNTIME=docker make update-with-container` (if you are using `podman`, set `RUNTIME=podman`).
 
 ## Verify
-Before you commit you changes, please run `make verify` locally to make sure you have generated all required files.
\ No newline at end of file
+Before you commit you changes, please run `make verify` locally to make sure you have generated all required files.
+
+## API upgrade flow
+OCM community doc [Changing the API](https://github.com/open-cluster-management-io/community/blob/main/sig-architecture/api_changes.md) describes the things to consider when updating the API, suggest you read it before upgrading the API.
+
+Once you decided to introduce a new version API, you will also need to consider deprecate the old version API and then remove it in the future. 
+Below is the suggested API migration flow, the example code is to migrate `ManagedClusterSet` API from v1beta1 to v1beta2.
+
+### Relase N (add new version)
+- Add new version v1beta2 API to api repo. For example:
+  - https://github.com/open-cluster-management-io/api/pull/174
+
+- Convert between old and new versions.
+
+  At this stage, both v1beta1 and v1beta2 are in served versions and storage version is v1beta1. If v1beta2 schema not change or can be converted to v1beta1, the CRD storage version can remain unchanged. If the v1beta2 schema changes and requires custom logic, you need to add conversion webhook to transform CRs between v1beta1 and v1beta2. For example: 
+  - https://github.com/open-cluster-management-io/registration/pull/272
+  - https://github.com/open-cluster-management-io/registration-operator/pull/279
+
+- The consumers (eg, ui/foundation/submarinar-addon/placement) CAN start to migrate to clusterset api v1beta2.
+
+- See also [How to add a new version](https://github.com/open-cluster-management-io/community/blob/main/sig-architecture/api_changes.md#how-to-add-a-new-version)
+
+### Relase N+1 (deprecated old version and start migration)
+- Mark v1beta1 as deprecated, served versions are v1beta1 and v1beta2, set storage: true on v1beta2.
+
```

Comment:
Do we consider to sperate: [set storage: true on v1beta2] and [migrate stored version of clusterset api to v1beta2] into 2 releases?

Release N+1: 
- Mark v1beta1 as deprecated, served versions are v1beta1 and v1beta2, set storage: true on v1beta2.

Release N+2:
- Migrate stored version of clusterset api to v1beta2 and remove v1beta1 from CRD `status.storedVersions`.

---

The motivation behind this is: if we done 2 things all in Release N+1, then we need to make sure storage set to v1beta2 happens first, then StorageVersionMigration created.
If any StorageVersionMigration is created before the storage: true move to v1beta2, the migration will end up with Succeeded condition but still using v1beta1 in the etcd.

Doing these 2 things in 2 releases can decouple the order dependency between 2 steps.

Reply:
@qiujian16  Please take another review, should we put this flow into best-practice?

Reply:
The risk we concerns here is this flow hasn't been verified by history. We never process in this way before, it's just a design.